### PR TITLE
BUG: Ensure series/frame mode() keeps int index

### LINKED ISSUE #33321: BUG: DataFrame.mode index dtype is not type stable
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.

- [x] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

The `.index.dtype` is not stable for DataFrame.mode. It depends on whether the DataFrame is empty and possible the dtypes

```python
In [60]: pd.DataFrame([], columns=['a', 'b']).mode().index.dtype
Out[60]: dtype('O')
```

#### Problem description

The index dtype should always be Int64, to match the non-empty case

#### Expected Output

```python
In [61]: pd.DataFrame({"A": ['a']}).mode().index.dtype
Out[61]: dtype('int64')
```

### PR DESCRIPTION
**Pull request by mzeitlin11**
- [x] closes #33321
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/core/frame.py [hunk 1]
```diff
@@ -9241,7 +9241,12 @@ def mode(
         def f(s):
             return s.mode(dropna=dropna)
 
-        return data.apply(f, axis=axis)
+        data = data.apply(f, axis=axis)
+        # Ensure index is type stable (should always use int index)
+        if data.empty:
+            data.index = Index([], dtype="int64")
+
+        return data
```
Comment (mzeitlin11) [inline pandas/core/frame.py:9249]: Empty result case doesn't hit `algorithms.mode`, so those changes aren't enough to ensure correct `dtype`

File: pandas/tests/frame/methods/test_mode.py [hunk 2]
```diff
@@ -0,0 +1,9 @@
+from pandas import DataFrame, Index
+import pandas._testing as tm
+
+
+def test_empty_df_mode():
+    df = DataFrame([], columns=["a", "b"])
+    result = df.mode()
+    expected = DataFrame([], columns=["a", "b"], index=Index([], dtype="int64"))
+    tm.assert_frame_equal(result, expected)
```
Comment (mzeitlin11) [inline pandas/tests/frame/methods/test_mode.py:9]: Couldn't find any `DataFrame.mode` testing - guessing it relies on correctness from `Series.mode()` testing (in `pandas/tests/test_algos.py::TestMode`). However, empty result case doesn't hit `Series.mode`, so some `DataFrame` testing seems necessary. Is there a better place for this test?

Comment (arw2019) [reply pandas/tests/frame/methods/test_mode.py:9]: We keep them in `pandas/tests/frame/test_reductions.py`

File: pandas/core/frame.py [hunk 3]
```diff
@@ -9241,7 +9241,12 @@ def mode(
         def f(s):
             return s.mode(dropna=dropna)
 
-        return data.apply(f, axis=axis)
+        data = data.apply(f, axis=axis)
+        # Ensure index is type stable (should always use int index)
+        if data.empty:
+            data.index = Index([], dtype="int64")
```
Comment (arw2019) [inline pandas/core/frame.py:9247]: is this right for 32-bit machines?

File: pandas/tests/frame/methods/test_mode.py (hunk 2 above)
Comment (mzeitlin11) [reply pandas/tests/frame/methods/test_mode.py]: Ah thanks! Sorry for missing that

File: pandas/core/frame.py (hunk 3 above)
Comment (mzeitlin11) [reply pandas/core/frame.py]: Changed to `int`, thanks for catching that

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/algorithms.py [hunk 4]
```diff
@@ -954,7 +954,9 @@ def mode(values, dropna: bool = True) -> Series:
         warn(f"Unable to sort modes: {err}")
 
     result = _reconstruct_data(result, original.dtype, original)
-    return Series(result)
+    # Ensure index is type stable (should always use int index)
+    index = None if len(result) else Index([], dtype=int)
```
[the remark this answers is not in the document]
Comment (mzeitlin11) [reply pandas/core/algorithms.py]: Done

**Review by jreback — APPROVED** [review]
lgtm ping on green

**Comment by mzeitlin11** [comment]
green

**Comment by jreback** [comment]
thanks @mzeitlin11

**Pull request merged**